README
You can install this using in the usual Perl fashion

	perl Makefile.PL
	make
	make test
	make install
	
The documentation is in the module file.  You can read the 
documentation with perldoc (in the same way before or after
you install it).

	perldoc scriptdist
	
This module is also in CVS on SourceForge

	http://sourceforge.net/projects/brian-d-foy/


Enjoy, 

brian d foy, bdfoy@cpan.org


Changes
Revision history for Perl module App::scriptdist

0.242 2017-04-24T05:44:38Z
	* Fix a testing problem on Windows

0.241 2017-03-29T05:29:44Z
	* Fix up various distro things. I neglected a package statement.
	Thus, this wasn't indexed. And then CPAN.pm went into a spiral
	in  https://rt.cpan.org/Ticket/Display.html?id=120781 [D

0.24 2015-08-17T22:04:15Z
	* Move some things into a module, App::scripdist
	* Everything is still really hacky.

0.23 - 2014-01-03
	* Get rid of MYMETA

0.22 - 2013-08-17
	* Update for Mastering Perl, 2e
	* Use git instead of CVS

0.21 - 2011-11-01
	* Check for minimum perl version
	* Extract prerequisites

0.16 - 2008-05-27
	* In compile.t template, use $^X, not "perl" (RT #36207)

0.14 - 2004-09-02
	* fixed a problem with the prereq.t test
	* cleaned up the distro
	* you don't need to upgrade if you already have this script

0.13 - 2004-02-22
	* don't look in CVS or Subversion directories for templates
	* minor fixes to default templates, mostly quoting issues

0.11 - 2004-02-07
    * use FindBin to get script name
    * some formatting fixes and warnings quashes
    * most fixes thanks to Soren Andersen

0.10 - 2004-01-12
	* this is the first release, a trial version




Makefile.PL
package App::scriptdist;
use strict;
use warnings;

=encoding utf8

=head1 The build file for App::scriptdist

This build file is a modulino; it works as both a build script and
a module.

To build the distribution, run this file normally:

	% perl Makefile.PL

But, it's more interesting than that. You can load it with C<require>
and call C<arguments> to get the data structure it passes to
C<WriteMakefile>:

	my $package = require '/path/to/Makefile.PL';
	my $arguments = $package->arguments;

Note that C<require>-ing a file makes an entry in C<%INC> for exactly
that name. If you try to C<require> another file with the same name,
even from a different path, C<require> thinks it has already loaded
the file. As such, I recommend you always require the full path to the
file.

The return value of the C<require> is a package name (in this case,
the name of the main module. Use that to call the C<arguments> method.

Even if this distribution needs a higher version of Perl, this bit
only needs v5.8. You can play with the data structure with a primitive
Perl.

=cut

use File::Spec::Functions qw(catfile);

my $module    = __PACKAGE__;
( my $dist = $module ) =~ s/::/-/g;

my $github    = 'https://github.com/briandfoy/scriptdist';
my $main_file = catfile( 'lib', split /::/, "$module.pm" );

my $script_name = 'scriptdist';

my %WriteMakefile = (
	'MIN_PERL_VERSION' => '5.010',
	'NORECURS'         => '1',

	'NAME'          => $module,
	'ABSTRACT_FROM' => $main_file,
	'VERSION_FROM'  => $main_file,
	'LICENSE'       => 'perl',
	'AUTHOR'        => 'brian d foy <bdfoy@cpan.org>',

	'EXE_FILES' =>  [ $script_name ],

	'CONFIGURE_REQUIRES' => {
		'ExtUtils::MakeMaker'   => '6.64',
		'File::Spec::Functions' => '0',
		},

	'BUILD_REQUIRES' => {
		},

	'TEST_REQUIRES' => {
		'Test::More' => '0.95',
		},

	'PREREQ_PM'     => {
		'File::Path' => '2.00',
		},

	'META_MERGE' => {
		'meta-spec' => { version => 2 },
		resources => {
			repository => {
				type => 'git',
				url  => "$github.git",
				web  => $github,
				},
			bugtracker => {
				web    => "$github/issues",
				},
			homepage => $github,
			},
		no_index => {
			dir       => [ qw( t ) ],
			directory => [ qw( t ) ],
			},
		},

	clean => { FILES => "$dist-* " .
		"t/test_home_dir/test_script.d" },
	);

1;

sub arguments { \%WriteMakefile }

do_it() unless caller;
sub do_it {
	require File::Spec;
	my $MM ='ExtUtils::MakeMaker';
	my $MM_version =
		eval{ "$MM " . $WriteMakefile{'CONFIGURE_REQUIRES'}{'ExtUtils::MakeMaker'} }
			||
		"$MM 6.64";
	eval "use $MM_version; 1" or die "Could not load $MM_version: $@";
	eval "use Test::Manifest 1.21"
		if -e File::Spec->catfile( qw(t test_manifest) );

	my $arguments = arguments();
	my $minimum_perl = $arguments->{MIN_PERL_VERSION} || '5.008';
	eval "require $minimum_perl;" or die $@;

	WriteMakefile( %$arguments );
	}


no warnings;
__PACKAGE__;


lib/App/scriptdist.pm
#!/usr/bin/perl
use utf8;
use v5.10;

package App::scriptdist;

use strict;
use warnings;

use vars qw(
	@EXPORT @EXPORT_OK %EXPORT_TAGS
	%Content $VERSION
	$Quiet
	);

$VERSION = '0.241';

@EXPORT_OK = qw( prompt find_files copy gitify content );
%EXPORT_TAGS = (
	'all' => [ @EXPORT_OK ],
	);

$Quiet = 0;

use Exporter qw(import);

use Cwd;
use ExtUtils::Command;
use ExtUtils::Manifest;
use File::Basename qw(basename);
use File::Find qw(find);
use File::Spec;
use FindBin ();

=encoding utf8

=head1 NAME

App::scriptdist - create a distribution around a perl script

=head1 SYNOPSIS

	use App::scriptdist qw(:all);

=head1 DESCRIPTION

This module provides the utility functions for the scriptdist program
that builds a basic Perl CPAN distribution around a standalone script
file that already exists.

I do not intend this for new development and don't want to create an
authoring tool. You can do that with some other tool (or fork this
one and build your own).

=head1 FUNCTIONS

=over 4

=item prompt( QUERY )

Provide a prompt, get the response, chomp the neewline, and return
the answer.

=cut

sub prompt {
	my( $query ) = shift;

	print $query;

	chomp( my $reply = <STDIN> );

	return $reply;
	}

=item find_files( DIRECTORY )

Find all the files under a directory.

=cut

sub find_files {
	my $directory = shift;

    my @files = ();

	my $wanted = sub {
		return unless -f $_;
		return if $_ =~ m<(?:CVS|\.svn|\.git)>;
		push @files, File::Spec->canonpath( $File::Find::name );
		};

    my %options = (
    	wanted => $wanted,
    	);

    find( \%options, $directory );

    return @files;
	}

=item copy( INPUT_FILE, OUTPUT_FILE, CONFIG_HASH )

Copy the file from one place to another.

=cut

sub copy {
	my( $input, $output, $hash ) = @_;

	print STDERR "Opening input [$input] for output [$output]\n";

	open my $in_fh,  '<', $input  or die "Could not open [$input]: $!\n";
	open my $out_fh, '>', $output or warn "Could not open [$output]: $!\n";

	my $count = 0;

	while( readline $in_fh ) {
		$count += s/%%SCRIPTDIST_(.*?)%%/$hash->{ lc $1 } || ''/gie;
		print {$out_fh} $_
		}

	print STDERR "Copied [$input] with $count replacements\n" unless $Quiet;
	}

=item gitify()

Unless the environment variable C<SCRIPTDIST_SKIP_GIT> is set, init
a git repo, add all the files, and make the initial commit.

=cut

sub gitify {
	return if $ENV{SCRIPTDIST_SKIP_GIT};
	chomp( my $git = `which git` );
	return unless length $git && -x $git;

	system $git, qw'init';
	system $git, qw'add .';
	system $git, qw'commit -a -m ', "Initial commit by $0 $VERSION";
	}

=item script_template( SCRIPT_NAME )

Return the script template.

=cut

sub script_template {
	my $script_name = shift;

	# Test::Pod thinks this stuff is pod if it's at the beginning
	# of the line
	my $script = <<"HERE";
	#!/usr/bin/perl

	=head1 NAME

	$script_name - this script does something

	=head1 SYNOPSIS

	=head1 DESCRIPTION

	=head1 AUTHOR

	=head1 COPYRIGHT

	=cut
HERE

	$script =~ s/^\s+//gm;

	return $script;
	}

=item content( CONFIG_HASH )

Return a hash reference of the contents of the files to add. The key
is the filename and the value is its contents.

=cut

sub content {
	my $hash = shift;

	$Content{"Changes"} =<<"CHANGES";
0.10 - @{ [ scalar localtime ] }
	+ initial distribution created with $hash->{name}
CHANGES

	$Content{"Makefile.PL"} =<<"MAKEFILE_PL";
use ExtUtils::MakeMaker 6.48;

eval "use Test::Manifest 1.21";

my \$script_name = "$$hash{script}";

WriteMakefile(
		'NAME'      => \$script_name,
		'VERSION'   => '$$hash{version}',

		'EXE_FILES' =>  [ \$script_name ],

		'PREREQ_PM' => {
@{ [
				join ",\n", (
					map ( {
						my $v = $_->version // 0;
						"\t\t\t" . $_->module . " => '$v'"
					} @{$hash->{modules}} ), ''
					),
 ] } 			},


		MIN_PERL_VERSION => $$hash{minimum_perl_version},

		clean => { FILES => "*.bak \$script_name-*" },
		);

1;
MAKEFILE_PL

	$Content{"MANIFEST.SKIP"} =<<"MANIFEST_SKIP";
#!include_default

\\.DS_Store
\\.releaserc
\\.svn
\\.git
$$hash{script}-.*

MANIFEST_SKIP

	$Content{".releaserc"} =<<"RELEASE_RC";
cpan_user @{[ $ENV{CPAN_USER} ? $ENV{CPAN_USER} : '' ]}
RELEASE_RC

	$Content{".gitignore"} =<<"GITIGNORE";
.DS_Store
.lwpcookies
$$hash{script}-*
blib
Makefile
pm_to_blib
GITIGNORE

	$Content{"t/test_manifest"} =<<"TEST_MANIFEST";
compile.t
pod.t
TEST_MANIFEST

	$Content{"t/pod.t"} = <<"POD_T";
use Test::More 0.98;
eval "use Test::Pod 1.00";
plan skip_all => "Test::Pod 1.00 required for testing POD" if \$@;
all_pod_files_ok();
POD_T

	$Content{"t/compile.t"} = <<"COMPILE_T";
use Test::More 0.98;

my \$file = "blib/script/$$hash{script}";

BAIL_OUT( "Script file is missing!" ) unless -e \$file;

my \$output = `$^X -c \$file 2>&1`;

BAIL_OUT( "Script file is missing!" ) unless
	like( \$output, qr/syntax OK\$/, 'script compiles' );

done_testing();
COMPILE_T

	\%Content;
	}

=back

=head1 TO DO

=over 4

=item * Copy modules into lib directory (to create module dist)

=item * Command line switches to turn things on and off

=back

=head2 Maybe a good idea, maybe not

=over 4

=item * Add a cover.t and pod coverage test?

=item * Interactive mode?

=back

=head1 SOURCE AVAILABILITY

This source is part of a Github project.

	https://github.com/briandfoy/scriptdist

=head1 CREDITS

Thanks to Soren Andersen for putting this script through its paces
and suggesting many changes to actually make it work.

=head1 AUTHOR

brian d foy, C<< <bdfoy@cpan.org> >>

=head1 COPYRIGHT

Copyright © 2004-2017, brian d foy C<< <bdfoy@cpan.org> >>. All rights reserved.

This code is available under the Artistic License 2.0.

=cut

1;
